src/legacy/core_plugins/embeddable_api/public/embeddables/embeddable.tsx

```diff
@@ -0,0 +1,185 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isEqual, cloneDeep } from 'lodash';
+import { Adapters } from 'ui/inspector';
+import * as Rx from 'rxjs';
+import { IContainer } from '../containers';
+import { IEmbeddable, EmbeddableInput, EmbeddableOutput } from './i_embeddable';
+import { ViewMode } from '../types';
+
+function getPanelTitle({
+  inputTitle,
+  defaultTitle,
+  hidePanelTitles,
+}: {
+  inputTitle?: string;
+  defaultTitle?: string;
+  hidePanelTitles?: boolean;
+}) {
+  if (hidePanelTitles) {
+    return '';
+  }
+
+  // Specificaly check for undefined, as an empty string override means "hide this title".
+  return inputTitle === undefined ? defaultTitle : inputTitle;
+}
+
+export abstract class Embeddable<
+  TEmbeddableInput extends EmbeddableInput = EmbeddableInput,
+  TEmbeddableOutput extends EmbeddableOutput = EmbeddableOutput
+> implements IEmbeddable<TEmbeddableInput, TEmbeddableOutput> {
+  public readonly parent?: IContainer;
+  public readonly isContainer: boolean = false;
+  public abstract readonly type: string;
+  public readonly id: string;
+
+  protected output: TEmbeddableOutput;
+  protected input: TEmbeddableInput;
+
+  private readonly input$: Rx.BehaviorSubject<TEmbeddableInput>;
+  private readonly output$: Rx.BehaviorSubject<TEmbeddableOutput>;
+
+  // Listener to parent changes, if this embeddable exists in a parent, in order
+  // to update input when the parent changes.
+  private parentSubscription?: Rx.Subscription;
+
+  private destoyed: boolean = false;
+
+  constructor(input: TEmbeddableInput, output: TEmbeddableOutput, parent?: IContainer) {
+    this.id = input.id;
+    this.output = {
+      title: getPanelTitle({
+        inputTitle: input.title,
+        defaultTitle: output.defaultTitle,
+        hidePanelTitles: input.hidePanelTitles,
+      }),
+      ...output,
+    };
+    this.input = {
+      viewMode: ViewMode.EDIT,
+      ...input,
+    };
+    this.parent = parent;
+
+    this.input$ = new Rx.BehaviorSubject<TEmbeddableInput>(this.input);
+    this.output$ = new Rx.BehaviorSubject<TEmbeddableOutput>(this.output);
+
+    if (parent) {
+      this.parentSubscription = Rx.merge(parent.getInput$(), parent.getOutput$()).subscribe(() => {
+        const newInput = parent.getInputForChild<TEmbeddableInput>(this.id);
+        this.onResetInput(newInput);
+      });
+    }
+  }
+
+  public abstract reload(): void;
+
+  public getInput$(): Readonly<Rx.Observable<TEmbeddableInput>> {
+    return this.input$.asObservable();
```

Comment:
@stacey-gammon because `.asObservable()` creates a new object; not just once, but every time you call `.getInput$()`.

Whereas type casting does nothing at runtime, it is free.